code changes to add a NuGet package recommender

birgithi opened the pull request:
Fixes: https://github.com/NuGet/Home/issues/9576

RElated https://github.com/nuget/home/issues/8924
Creating a PR to get feedback on code changes to add a NuGet package recommender

Here's an example of what it currently looks like. The current icon will be replaced by the star.
![Capture](https://user-images.githubusercontent.com/5205104/77108121-e989cd80-69de-11ea-8ddc-d9b15f4fa3db.PNG)

I updated t;he PR with a newer icon. This replaces the rotated square with a start in front of the package id:
![Capture2](https://user-images.githubusercontent.com/5205104/77485508-8aff8d80-6dea-11ea-8c41-958a6bbaee8f.PNG)

birgithi committed:
Initial checking of recommender changes

birgithi committed:
Initial checking of recommender changes

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/signed)](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=3299) <br/>All CLA requirements met.

zivkan commented:
@birgithi Can you please create an issue at https://github.com/NuGet/Home/issues/new, if there isn't one already, and then edit the first comment in the PR and put a link to the issue in there?  We use this for tracking and release notes.  Thanks.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageIconMonikers.cs:
```diff
@@ -12,12 +12,14 @@ public static class PackageIconMonikers
         public static ImageMoniker DownloadIndicator => new ImageMoniker { Guid = ManifestGuid, Id = _downloadIndicator };
         public static ImageMoniker InstalledIndicator => new ImageMoniker { Guid = ManifestGuid, Id = _installedIndicator };
         public static ImageMoniker MigratorHelpIcon => new ImageMoniker { Guid = ManifestGuid, Id = _migratorHelpIcon };
+        public static ImageMoniker RecommendedIndicator => new ImageMoniker { Guid = ManifestGuid, Id = _recommendedIndicator };
```
Can you post a screenshot in the PR description so we can see how everything looks like?

birgithi replied:
Add a screenshot.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -20,8 +20,9 @@ internal class PackageItemLoader : IPackageItemLoader
         private readonly bool _includePrerelease;
 
-        private readonly IPackageFeed _packageFeed;
+        private readonly IEnumerable<IPackageFeed> _packageFeeds;
```
If we are always gonna have 2 package feed, I'd just call them out specifically. 
No need for a collection.

birgithi replied:
Changed this to always have two package feeds, the main feed and the recommender feed.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -20,8 +20,9 @@ internal class PackageItemLoader : IPackageItemLoader
-        private readonly IPackageFeed _packageFeed;
+        private readonly IEnumerable<IPackageFeed> _packageFeeds;
         private PackageCollection _installedPackages;
+        private List<string> _recommendedPackages;
```
Use interfaces.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -34,7 +35,7 @@ internal class PackageItemLoader : IPackageItemLoader
         public IItemLoaderState State => _state;
 
-        public bool IsMultiSource => _packageFeed.IsMultiSource;
+        public bool IsMultiSource => _packageFeeds.Last().IsMultiSource;
```
The above feedback makes this moot.

birgithi replied:
Removed this change since we now always have two package feeds, this main one and the recommender feed.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -50,6 +53,80 @@ internal class PackageLoadContext
+            List<LockFile> lockFiles = new List<LockFile>();
+            var lockFileFormat = new LockFileFormat();
+
+            foreach (var project in buildIntegratedProjects)
```
We probably need an abstraction on NuGetProject for this. 

We should also consider paralleling this work.

nkolev92 replied:
I also could be wrong, but trying to roughly follow the code path this is happening on the UI thread, and will cause UI delays.

zivkan replied:
+1.  Is this intentional to not support `packages.config` projects?

birgithi replied:
I've removed the code that gets dependency packages for PR-style projects from the assets file. We agreed that for right now we would only make recommendations for PC-style projects while we find a better way to get the dependency packages for PR-style projects.

nkolev92 replied:
GetSupportedFrameworks is the method currently doing it for PC, so we can make improve that.

birgithi replied:
I removed the added GetTargetFrameworksAsync method since it was only added to support PR-style projects. I agree it would be better to improve the existing GetSupportedFrameworks method to add support for PR-style projects.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -50,6 +53,80 @@ internal class PackageLoadContext
+        }
+
+        // get target frameworks for BuildIntegratedNuGetProjects
+        public async Task<IEnumerable<string>> GetTargetFrameworksAsync()
```
We have some code duplication here. 
The assets file contains the frameworks as well. If we have to read the assets file then we can avoid calling GetPackageSpecs. 

In general moving the abstraction to NuGet project probably works better. 

@rrelyea We discussed this in the past, but we should consider having a NuGet engineering add this abstraction or at the least work with @birgithi to help her navigate all the craziness we have.

birgithi replied:
I re;moved this method here and added one to NuGetPackageManager.cs instead to get target frameworks. Let me know if there's a better way to do this.

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs:
```diff
@@ -34,7 +34,7 @@ public override async Task<IEnumerable<IPackageSearchMetadata>> SearchAsync(stri
         }
 
-        private static IEnumerable<VersionInfo> GetVersions(PackageSearchMetadata metadata, SearchFilter filter)
+        protected static IEnumerable<VersionInfo> GetVersions(PackageSearchMetadata metadata, SearchFilter filter)
```
I don't imagine you need this change anymore.
You can revert it by running the following from the command line

```console
git checkout dev -- src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs 
```

birgithi replied:
Thanks. I reverted the change.

birgithi committed:
undo change to GetVersions. This can stay private.

birgithi committed:
undo change to GetVersions. This can stay private.

birgithi committed:
Changed recommended package icon to star character before package id
Changed package feeds list to two package feeds, main and recommender
Replaced simulated recommendations by call to VS recommender service
Added IsRecommended to IPackageSearchMetadata
General code cleanup

birgithi committed:
change square icon to star

birgithi committed:
Updated unit tests to match code changes

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/Model/PackageSearchMetadata.cs:
```diff
@@ -199,6 +199,9 @@ private static IList<string> GetNonStandardLicenseIdentifiers(NuGetLicenseExpres
         public bool IsListed { get; private set; } = true;
 
+        [JsonIgnore]
+        public bool IsRecommended { get; set; } = false;
```
Could leave off the default value since false is implicit.

birgithi replied:
Removed default value.

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs:
```diff
@@ -70,6 +70,8 @@ public string Tags
 
         public string Title => !string.IsNullOrEmpty(_nuspec.GetTitle()) ? _nuspec.GetTitle() : _nuspec.GetId();
 
+        public bool IsRecommended { get; set; } = false;
```
Could leave off the default value since false is implicit.

birgithi replied:
Removed default value.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -81,6 +83,7 @@ public LoadingStatus LoadingStatus
         public PackageItemLoader(
             PackageLoadContext context,
             IPackageFeed packageFeed,
+            IPackageFeed recommenderPackageFeed,
```
Perhaps make this optional with `= null` as a default to avoid passing `null` from several places.

birgithi replied:
Made recommenderPackageFeed an optional parameter

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageItemLoaderTests.cs:
```diff
@@ -45,7 +45,7 @@ public async Task MultipleSources_Works()
 
             var packageFeed = new MultiSourcePackageFeed(repositories, logger: null, telemetryService: null);
-            var loader = new PackageItemLoader(context, packageFeed, "nuget");
+            var loader = new PackageItemLoader(context, packageFeed, null, "nuget");
```
If you don't use the default value in the constructor, please use a named parameter here (`recommenderPackageFeed: null`).

birgithi replied:
Reverted this change since I made recommenderPackageFeed an optional parameter.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageItemLoaderTests.cs:
```diff
@@ -123,7 +123,7 @@ public async Task EmitsSearchTelemetryEvents()
 
                 // Act
-                var loader = new PackageItemLoader(context, packageFeed, searchText: "nuget", includePrerelease: true);
+                var loader = new PackageItemLoader(context, packageFeed, null, searchText: "nuget", includePrerelease: true);
```
If you don't use the default value in the constructor, please use a named parameter here (`recommenderPackageFeed: null`).

birgithi replied:
Reverted this change since I made recommenderPackageFeed an optional parameter.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageItemLoaderTests.cs:
```diff
@@ -182,7 +182,7 @@ public async Task GetTotalCountAsync_Works()
 
             var packageFeed = new MultiSourcePackageFeed(repositories, logger: null, telemetryService: null);
-            var loader = new PackageItemLoader(context, packageFeed, "nuget");
+            var loader = new PackageItemLoader(context, packageFeed, null, "nuget");
```
If you don't use the default value in the constructor, please use a named parameter here (`recommenderPackageFeed: null`).

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageItemLoaderTests.cs:
```diff
@@ -200,7 +200,7 @@ public async Task LoadNextAsync_Works()
             var context = new PackageLoadContext(null, false, uiContext);
             var packageFeed = new TestPackageFeed();
-            var loader = new PackageItemLoader(context, packageFeed, TestSearchTerm, true);
+            var loader = new PackageItemLoader(context, packageFeed, null, TestSearchTerm, true);
```
If you don't use the default value in the constructor, please use a named parameter here (`recommenderPackageFeed: null`).

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageItemLoaderTests.cs:
```diff
@@ -256,7 +256,7 @@ public async Task PackageReader_NotNull()
 
                 var packageFeed = new MultiSourcePackageFeed(repositories, logger: null, telemetryService: null);
-                var loader = new PackageItemLoader(context, packageFeed, "nuget");
+                var loader = new PackageItemLoader(context, packageFeed, null, "nuget");
```
If you don't use the default value in the constructor, please use a named parameter here (`recommenderPackageFeed: null`).

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageManagerListItemsTest.cs:
```diff
@@ -43,7 +43,7 @@ public async Task PackagePrefixReservation_FromOneSource()
 
             var packageFeed = new MultiSourcePackageFeed(repositories, logger: null, telemetryService: null);
-            var loader = new PackageItemLoader(context, packageFeed, "EntityFramework", false);
+            var loader = new PackageItemLoader(context, packageFeed, null, "EntityFramework", false);
```
If you don't use the default value in the constructor, please use a named parameter here (`recommenderPackageFeed: null`).

birgithi replied:
Reverted this change since I made recommenderPackageFeed an optional parameter.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageManagerListItemsTest.cs:
```diff
@@ -101,7 +101,7 @@ public async Task PackagePrefixReservation_FromMultiSource()
 
             var packageFeed = new MultiSourcePackageFeed(repositories, logger: null, telemetryService: null);
-            var loader = new PackageItemLoader(context, packageFeed, "EntityFramework", false);
+            var loader = new PackageItemLoader(context, packageFeed, null, "EntityFramework", false);
```
If you don't use the default value in the constructor, please use a named parameter here (`recommenderPackageFeed: null`).

birgithi replied:
Reverted this change since I made recommenderPackageFeed an optional parameter.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/TestPackageSearchMetadata.cs:
```diff
@@ -54,6 +54,8 @@ class TestPackageSearchMetadata : IPackageSearchMetadata
 
         public LicenseMetadata LicenseMetadata { get; set; }
 
+        public bool IsRecommended { get; set; } = false;
```
Could leave off the default value since false is implicit.

birgithi replied:
Removed default value.

donnie-msft commented on src/NuGet.Clients/NuGet.Tools/Properties/AssemblyInfo.cs:
```diff
@@ -13,6 +13,7 @@
 [assembly: ProvideCodeBase(CodeBase = @"$PackageFolder$\Lucene.Net.dll")]
 [assembly: ProvideCodeBase(CodeBase = @"$PackageFolder$\Microsoft.Web.XmlTransform.dll")]
 [assembly: ProvideCodeBase(CodeBase = @"$PackageFolder$\Newtonsoft.Json.dll")]
+[assembly: ProvideCodeBase(CodeBase = @"$PackageFolder$\NugetRecommender.VisualStudio.Contracts.dll")]
```
Please change spelling to: NuGet

birgithi replied:
Removed here, but fixed spelling in other locations.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,205 @@
+        private readonly IPackageMetadataProvider _metadataProvider;
+        private readonly Common.ILogger _logger;
+
+        Recommender.IVsNugetPackageRecommender NugetRecommender { get; set; }
```
Please change spelling to: NuGet

birgithi replied:
Changed spelling to NuGet

donnie-msft reviewed (changes requested):
Thanks for your contribution! Few minor suggestion from me...

birgithi committed:
Added real recommender and contracts package. Added telemetry for selected index, num recommended and target framework.

birgithi committed:
Changes in response to PR feedback

birgithi committed:
Added RecommendPackage telemetry property. Added A/B testing.

birgithi committed:
Added RecommendPackages telemetry event

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -563,6 +586,9 @@ List<TelemetryEvent> ToTelemetryPackageList(List<Tuple<string, string>> packages
                 // userAction.Version can be null for deleted packages.
                 actionTelemetryEvent.ComplexData["SelectedPackage"] = ToTelemetryPackage(new Tuple<string, string>(userAction.PackageId, userAction.Version?.ToNormalizedString() ?? string.Empty));
+                actionTelemetryEvent["SelectedIndex"] = selectedIndex;
+                actionTelemetryEvent["NumRecommended"] = numRecommended;
```
```suggestion
                actionTelemetryEvent["RecommendedCount"] = numRecommended;
```

It's in line with the rest of our metrics.

birgithi replied:
Changed to RecommendedCount

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -613,6 +639,12 @@ List<TelemetryEvent> ToTelemetryPackageList(List<Tuple<string, string>> packages
             }
+
+            // target framworks
+            if (targetFrameworks != null && targetFrameworks.Count > 0)
```
nit suggestion.

```suggestion
            if (targetFrameworks?.Count > 0)
```

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:
```diff
@@ -65,6 +65,37 @@ public virtual void CleanUp()
         /// </summary>
         public abstract IEnumerable<NuGetProject> GetSelectedProjects(UserAction action);
 
+        private int _selectedIndex = 0;
```
Use auto property here if you need to expose it publicly.

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:
```diff
@@ -65,6 +65,37 @@ public virtual void CleanUp()
+        {
+            get { return _selectedIndex; }
+        }
+        public int NumRecommended
```
style: empty lines between method declarations.

birgithi replied:
No longer applies after above fix.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -192,6 +196,42 @@ public async Task UpdateStateAsync(IProgress<IItemLoaderState> progress, Cancell
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            if (_packageFeed == null)
```
You don't need this. 
There's a null check in the constructor of this class.

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -192,6 +196,42 @@ public async Task UpdateStateAsync(IProgress<IItemLoaderState> progress, Cancell
         }
 
+        private async Task<SearchResult<IPackageSearchMetadata>> CombineSearchAsync(
+            ContinuationToken continuationToken,
```
You are not using the continuation token

birgithi replied:
Sorry, missed this comment earlier. This method has been updated since then and uses the continuation token.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -50,6 +53,62 @@ internal class PackageLoadContext
+                var lockFilePath = await project.GetAssetsFilePathAsync();
+                if (File.Exists(lockFilePath))
+                {
+                    lockFiles = lockFiles.Append(lockFileFormat.Read(lockFilePath)).ToList();
```
This incurs a very expensive performance hit. 
I'm not sure how I'd want us to do this yet, but we should consider caching it, at least for the context of the PM UI.

zivkan replied:
Are transitive dependencies important to the package recommender, or only the direct dependencies?  Because the direct dependencies will already be cached in the project cache's PackageSpec (or is it a DGSpec?), so no need to load the assets file from disk

birgithi replied:
Yes, transitive dependencies are used by the recommender. We don't want to recommend a package that is already included as a dependency.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -50,6 +53,62 @@ internal class PackageLoadContext
+        {
+            // get lock files from projects. This will add all dependency packages for all projects in the list
+            List<LockFile> lockFiles = await getLockFiles(Projects.ToList());
+            List<PackageCollectionItem> dependentPackages = new List<PackageCollectionItem>();
```
I'm not sure why you need package collection, it's too heavy of a model.

birgithi replied:
Changed this to a dictionary containing package id and version range.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -50,6 +53,62 @@ internal class PackageLoadContext
+                    {
+                        foreach (Packaging.Core.PackageDependency dep in lib.Dependencies)
+                        {
+                            Versioning.NuGetVersion version;
```
Why do we need only one version?
We shouldn't use Versioning.NuGetVersion, we should do the correct using just refer to it as NuGetVersion.
We only use namespaces when there are naming clashes.

birgithi replied:
Now returning version range. 
Added using for Versioning.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/INuGetUI.cs:
```diff
@@ -83,6 +83,10 @@ public interface INuGetUI
 
+        int SelectedIndex { get; }
+        int NumRecommended { get; }
+        bool RecommendPackages { get; }
```
Logically, if feels weird that these are defined in an interface.

nkolev92 replied:
Similar feedback as above about the spacing and naming.

birgithi replied:
Removed these from the interface. But they are still in NuGetUI.cs. Should I try to find a way to remove them there as well?

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -98,6 +101,13 @@ internal PackageSourceMoniker SelectedSource
         private readonly Stopwatch _sinceLastRefresh;
 
+        // Helper class to pass back multiple package feeds
+        private class PackageFeeds
```
You can use a value tuple instead of a class:
https://docs.microsoft.com/en-us/dotnet/api/system.valuetuple?view=netframework-4.8

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -719,6 +729,12 @@ private void SearchPackagesAndRefreshUpdateCount(bool useCacheForUpdates)
+        public bool IsRecommenderFlightEnabled()
+        {
+            // TODO: Need to use correct flight name
+            return ExperimentationService.Default.IsCachedFlightEnabled("nugetrecommendpkgs");
```
Does this make a network call? 
Any threading affinity?

nkolev92 replied:
This is currently happening on the UI thread and that can be problematic.

birgithi replied:
It doesn’t make a network call.  It reads information from local storage and is intended to be fast and cheap.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -733,10 +749,37 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
+            //   this is the Browse tab,
+            //   and the search text is an empty string
+            if (loadContext.SourceRepositories.Count() == 1
+                && loadContext.SourceRepositories.First().PackageSource.Source.ToLower() == "https://api.nuget.org/v3/index.json"
```
There should be a utility that already does this. 
Search for "IsNuGetOrg" or something like that.

birgithi replied:
I changed it to TelemetryUtility.IsNuGetOrg. But should we be using TelemetryUtility for this? I couldn't find any other method that does this.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -733,10 +749,37 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
+            //   and the search text is an empty string
+            if (loadContext.SourceRepositories.Count() == 1
+                && loadContext.SourceRepositories.First().PackageSource.Source.ToLower() == "https://api.nuget.org/v3/index.json"
+                && loadContext.IsSolution == false
```
I would switch the ordering here. 
You want the cheapest check to be first.

birgithi replied:
Fixed

zivkan replied:
`.Count()` is a LINQ method that has to iterate an `IEnumerable` (or use reflection to figure out if it's actually a list that has a `.Count` property). All the other conditions checking with `==` are properties that do not require calculations, so are cheaper.

birgithi replied:
Moved the .Count() condition after all of the other == conditions.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -733,10 +749,37 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
+                && loadContext.SourceRepositories.First().PackageSource.Source.ToLower() == "https://api.nuget.org/v3/index.json"
+                && loadContext.IsSolution == false
+                && _topPanel.Filter == ItemFilter.All
+                && searchText == "")
```
Use `string.Empty`

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -1014,7 +1094,7 @@ private void Filter_SelectionChanged(object sender, FilterChangedEventArgs e)
         private void Refresh()
         {
-            if (_topPanel.Filter != ItemFilter.All)
+            if (_topPanel.Filter != ItemFilter.All || _recommendPackages)
```
Can you elaborate why we need this check? 

As is, we'd refresh a lot more often than we should.

birgithi replied:
If a user installs or uninstalls a package, that would change what packages we recommend. So we would want this list to be refreshed if the recommender is being used.

nkolev92 replied:
That's not the only case where this code path runs. I think the comment is misleading. 

If you look at the invocations of that method there sohuld be like 9 or different code paths calling into it. 

This check was basically "if on install/updates/consolidate" tab, refresh the whole list...but now we are refreshing when on the browse tab too even if nothing has changed.

birgithi replied:
I've removed this for now while I look for a more targeted way to trigger a refresh.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,226 @@
+
+        public RecommenderPackageFeed(
+            SourceRepository sourceRepository,
+            IEnumerable<PackageCollectionItem> installedPackages,
```
The packages can change during the life time of a package manager UI. 
So that'd imply that the recommendations should change as well.

birgithi replied:
That was my intent when I added the check for recommendations in the Refresh method. Are you asking for a different change here? Can you clarify?

nkolev92 replied:
Does this collection get updated when the installed packages change?

donnie-msft replied:
I think what happens is the Search logic creates a new `RecommenderPackageFeed`. This is relying on the fact that Installing a package triggers a refresh of the PMUI, which triggers the aforementioned Search logic.

birgithi replied:
I looked at the different ways RecommendPackagesAsync could be called, and didn't find any path where the installed packages could have changed without creating a new RecommenderPackageFeed.

nkolev92 replied:
Sounds good. Resolving.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,226 @@
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine(ex);
```
I'd avoid debugs like this. 
If important ensure that they are communicated in an actionable manner. 
Ideally the recommender doesn't throw. 

Is throwing part of the contract?

birgithi replied:
Removed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/RecommendTelemetryEvent.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public RecommendTelemetryEvent(int numRecommended, double duration) : base("RecommendPackages")
+        {
+            base["NumRecommended"] = numRecommended;
```
Count...

birgithi replied:
Moved this telemetry event to the recommender.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/RecommendTelemetryEvent.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class RecommendTelemetryEvent : TelemetryEvent
+    {
+        public RecommendTelemetryEvent(int numRecommended, double duration) : base("RecommendPackages")
```
This feels like a very specific tel event. Can we attach it to some existing events if possible?

Ideally the recommender itself can measure its performance too.

birgithi replied:
Moved this telemetry event to the recommender.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude:
```diff
@@ -30,6 +30,7 @@ NuGet.ProjectModel.dll
 NuGet.Protocol.dll
 NuGet.Resolver.dll
 NuGet.Versioning.dll
+Microsoft.DataAI.NuGetRecommender.Contracts.dll
```
Isn't this assembly supposed to be embeddable?
Might need to add it embedinterops.targets and move this to vsixexlude.

birgithi replied:
I tried adding it to embedinterops.targets and moving this from .vsixinclude to .vsixignore but I get errors :
Cannot embed interop types from assembly 'Microsoft.DataAI.NuGetRecommender.Contracts, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' because it is missing the 'System.Runtime.InteropServices.GuidAttribute' attribute.
Cannot embed interop types from assembly 'Microsoft.DataAI.NuGetRecommender.Contracts, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' because it is missing either the 'System.Runtime.InteropServices.ImportedFromTypeLibAttribute' attribute or the 'System.Runtime.InteropServices.PrimaryInteropAssemblyAttribute' attribute.
Am I missing another step?

nkolev92 replied:
Looks like the assembly was not built as embeddable. cc @bmajik 

Is the assembly embeddable or?

birgithi replied:
Fixed now.

nkolev92 commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -3392,5 +3392,29 @@ private RestoreCommandProvidersCache GetRestoreProviderCache()
+                frameworks.Add(framework.ToString());
+            }
+
+            if (project is BuildIntegratedNuGetProject)
```
You can use `as` to check and assign to a new variable.

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -3392,5 +3392,29 @@ private RestoreCommandProvidersCache GetRestoreProviderCache()
+                var packageSpecs = await biproject.GetPackageSpecsAsync(dgcContext);
+                foreach (var packageSpec in packageSpecs)
+                {
+                    frameworks = frameworks.Concat(packageSpec.TargetFrameworks.Select(tf => tf.FrameworkName.ToString())).ToHashSet();
```
Why do we need a hash set? 

We can probably add an interface method to get the frameworks for Build Integrated projects.

cc @zivkan thoughts?

birgithi replied:
Changed to a list.

zivkan replied:
sorry, didn't see this mention for an entire week :(

Regarding hashset vs list, it's complicated :)  if it's a performance question, we'd really need to do a microbenchmark to be sure. `HashSet` has overheads, so I've often read that when the collection is small, it's faster to use `List`. However, semantically how we're using it, to find a unique collection of TFMs, `HashSet` is the perfect fit conceptually. Having said that, adding new items to the `HashSet` by using `.Concat`, `.Select` and `.ToHashSet` is just about the worst performance without intentionally writing slow code. If HashSet doesn't have a `AddRange`, iterate `packageSpec.TargetFrameworks` and `.Add` one at a time. That's what `.ToHashSet` does internally.

I don't fully understand the comment about adding an interface method for build integrated projects. Are you suggesting moving this method to `BuildIntegratedNuGetProject`?  If we plan to support package recommendations for `PackageReference` projects only, then that makes sense.

nkolev92 replied:
> I don't fully understand the comment about adding an interface method for build integrated projects. Are you suggesting moving this method to BuildIntegratedNuGetProject? If we plan to support package recommendations for PackageReference projects only, then that makes sense.

There is a way to get the target framework in PC projects already. 
What makes build integrated projects special is the fact that they can multi target.

birgithi replied:
Is there an additional change I should make here? Or is this okay now?

nkolev92 replied:
This method is now complete removed, so I imagine it's ok (I did not look, but inferred from the other comments).

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Model/IPackageSearchMetadata.cs:
```diff
@@ -31,6 +31,7 @@ public interface IPackageSearchMetadata
         string Summary { get; }
         string Tags { get; }
         string Title { get; }
+        bool IsRecommended { get; set; }
```
This is incorrect. 

We should not change any NuGet.Core/* types for this change. 

This type maps to a server model, and recommendations are not related to servers.

birgithi replied:
Removed.

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Model/LocalPackageSearchMetadata.cs:
```diff
@@ -70,6 +70,8 @@ public string Tags
 
         public string Title => !string.IsNullOrEmpty(_nuspec.GetTitle()) ? _nuspec.GetTitle() : _nuspec.GetId();
 
+        public bool IsRecommended { get; set; }
```
Same as the previous ones.

birgithi replied:
Removed.

birgithi committed:
Added contracts dll to assembly info

birgithi committed:
Change NumRecommended property to RecommendedCount

birgithi committed:
Code cleanup based on feedback

birgithi committed:
More code cleanup

birgithi committed:
Remove IsRecommended from NuGet.Core and properties from INuGetUI

birgithi committed:
Code cleanup to use simpler types

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -3392,5 +3392,35 @@ private RestoreCommandProvidersCache GetRestoreProviderCache()
             return _restoreProviderCache;
         }
+
+        public static async Task<List<string>> GetTargetFramework(NuGetProject project)
```
There's already a similar helper method [`AssemblyLoadContext.GetSupportedFrameworks`](https://github.com/NuGet/NuGet.Client/blob/1fbc87841107a627c6f4349f733e48c28fcff028/src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs#L54-L100), which is used in PM UI.

If that helper isn't suitable, please move this helper to somewhere else in PM UI's code. Any code under `src/NuGet.Core` basically becomes part of the NuGet SDK which we need to support and not break backwards compatibility.

birgithi replied:
Moved this to package load context.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,201 @@
+                _telemetryService.EmitTelemetryEvent(new RecommendTelemetryEvent(
+                // this is the total number of package ids returned from the recommender, not the number actually recommended
+                recommendIds.Count(),
+                duration.TotalSeconds));
```
Is this telemetry that belongs in the recommender service, not NuGet?

birgithi replied:
Moved this telemetry to the recommender service.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -512,7 +518,21 @@ private static string GenerateUpgradeReport(NuGetProject nuGetProject, string ba
+                        removedPackages,
+                        updatedPackagesOld,
+                        updatedPackagesNew,
+                        frameworks.ToList());
```
`frameworks` is already a list. Calling `ToList` uses extra CPU cycles and causes extra memory allocations (it might be ok when you want immutable or readonly lists, but then `AddUiActionEngineTelemetryProperties`'s input parmaeter should reflect that. But even then, `frameworks` is not kept or modified anywhere else, so there's no need for an extra copy).

birgithi replied:
Fixed. (I missed updating this after changing frameworks to a List)

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:
```diff
@@ -65,6 +65,23 @@ public virtual void CleanUp()
         /// </summary>
         public abstract IEnumerable<NuGetProject> GetSelectedProjects(UserAction action);
 
+        public int SelectedIndex { private set;  get; }
+        public int RecommendedCount { private set;  get; }
+        public bool RecommendPackages { private set;  get; }
```
getters are usually defined before setters, but I can't find any coding guidelines about this, so I don't know if it needs to change. Just feels odd to me.

birgithi replied:
Changed order.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -50,6 +55,48 @@ internal class PackageLoadContext
         public Task<PackageCollection> GetInstalledPackagesAsync() =>_installedPackagesTask;
 
+        // get lock file for projects
+        public async Task<List<LockFile>> getLockFiles(IEnumerable<NuGetProject> nuGetProjects)
```
This method is only called from within the class, so should be private.  Also, method names should start with a capital.

birgithi replied:
Made this private and capitalized the name.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx:
```diff
@@ -315,6 +315,9 @@
     <value>Latest stable version is installed</value>
   </data>
+  <data name="ToolTip_PackageRecommended" xml:space="preserve">
+    <value>Recommended package</value>
```
You've probably already had discussions about this, but I have issues with the word "recommend". A [dictionary definition of the word](https://www.bing.com/search?q=define+recommend&cvid=c93157fdac504e1ea5d879b320f30b5d&FORM=ANNTA1&PC=U531#) says that it's something we're giving our approval for, which implies we do not give our approval of other packages.  I think "Suggested package" implies less favouritism.

birgithi replied:
This is more of a PM question. The UI mockup from Karan has "IntelliCode package suggestion based on this project context", so I changed it to that.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,201 @@
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+using Microsoft.VisualStudio.Shell;
+using Recommender = Microsoft.DataAI.NuGetRecommender.Contracts;
```
Is the `Recommender` alias needed (is there ambiguous types without it)?

In any case, this using statement is not sorted either alias or by namespace. We very, very rarely use aliases, but my memory is we usually put aliases after the not-aliased namespace usings.

birgithi replied:
Removed the alias and sorted the usings.

birgithi committed:
More code cleanup

birgithi committed:
removed RecommendPackages telemetry. Move GetTargetFrameworks from Core to package load context.

birgithi committed:
Remove code that gets dependency packages from the assets file. Limit recommender to only PC-style projects.

birgithi committed:
Add recommender version info

birgithi committed:
embed contracts assembly

birgithi committed:
Update version of recommender contracts package

nkolev92 commented on src/NuGet.Clients/NuGet.Tools/Properties/AssemblyInfo.cs:
```diff
@@ -10,6 +10,7 @@
 [assembly: AssemblyDescription("Visual Studio Extensibility Package (vsix)")]
 [assembly: ComVisible(false)]
 
+[assembly: ProvideCodeBase(CodeBase = @"$PackageFolder$\Microsoft.DataAI.NuGetRecommender.Contracts.dll")]
```
Should be unnecessary right now, because it's embedded

birgithi replied:
Removed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -512,7 +515,22 @@ private static string GenerateUpgradeReport(NuGetProject nuGetProject, string ba
+                        userAction,
+                        nuGetUI?.SelectedIndex,
+                        nuGetUI?.RecommendedCount,
+                        nuGetUI?.RecommendPackages,
```
I imagine we are looking for `RecommendedPackages` instead here?

birgithi replied:
Actually, this is a boolean indicating that this is a situation where we would have asked for recommendations independent of A/B testing status. For example whether or not there is a search string or if the package source is nuget.org. This is so that we can make correct comparisons for the A/B tests.

nkolev92 replied:
Ahh...makes sense.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -50,6 +55,37 @@ internal class PackageLoadContext
 
         public Task<PackageCollection> GetInstalledPackagesAsync() =>_installedPackagesTask;
 
+        public async Task<List<string>> GetTargetFrameworksAsync()
```
I think we can remove this and not do it in V1 because we are not doing PR in v1.

birgithi replied:
Removed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -184,6 +191,16 @@ internal PackageSourceMoniker SelectedSource
+            {
+                _forceRecommender = (Environment.GetEnvironmentVariable("RecommendNuGetPackages") == "1");
+            }
+            catch(SecurityException)
```
nit: formatting.

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -733,10 +755,45 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
+            // Check for A/B experiment here. For control group, call CreatePackageFeedAsync with false instead of _recommendPackages
+            // Also check if the environment variable RecommendNuGetPackages is set to 1
+            var packageFeeds = (mainFeed: (IPackageFeed)null, recommenderFeed: (IPackageFeed)null);
+            if (IsRecommenderFlightEnabled() || _forceRecommender)
```
Maybe just capture it all in one method?

birgithi replied:
I moved the _forceRecommender check inside IsRecommenderFlightEnabled(). Is that what you meant?

nkolev92 replied:
Yes.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,192 @@
+            _logger = logger;
+
+            // Get NuGet package recommender service, or null if it is not available
+            NuGetRecommender = Package.GetGlobalService(typeof(SVsNuGetRecommenderService)) as IVsNuGetPackageRecommender;
```
This needs the UI thread. 
We should not use thread affinitized methods in constructors. 

This should be either loaded through MEF/lazily.

birgithi replied:
Working on this. Will let you know when I have a fix.

birgithi replied:
I updated this to use AsyncLazy

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,192 @@
+
+            // Get NuGet package recommender service, or null if it is not available
+            NuGetRecommender = Package.GetGlobalService(typeof(SVsNuGetRecommenderService)) as IVsNuGetPackageRecommender;
+            if(NuGetRecommender != null)
```
style: fomratting

birgithi replied:
Fixed

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/INuGetUI.cs:
```diff
@@ -109,4 +109,4 @@ public interface INuGetUI
     }
-}
\ No newline at end of file
+}
```
Revert this sicne it's an unnecessary change.

birgithi replied:
VS automatically adds a newline at the end of each file. Should I remove them from other files as well? Or just this one since it's the only change to this file?

nkolev92 replied:
Just this one because it's the only change.

nkolev92 reviewed (commented):
Left a few comments to address.

Ensure all the nuget code is formatted correctly. 
Ensure that the formatting does not affect too older formatting in the file and create a super large change set.

nkolev92 commented:
@donnie-msft @zivkan 

We'd appreciate a review here.

We want to merge this to a feature branch first and then create follow up issues for the clean-up.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -502,6 +502,9 @@ private static string GenerateUpgradeReport(NuGetProject nuGetProject, string ba
                         }
                     }
 
+                    PackageLoadContext plc = new PackageLoadContext(null, false, uiService.UIContext);
```
Please use named parameters for null and false.

birgithi replied:
Fixed

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageItemControl.xaml:
```diff
@@ -149,6 +149,15 @@
+                        ToolTip="{x:Static nuget:Resources.ToolTip_PackageRecommended}"
+                        AutomationProperties.Name="{x:Static nuget:Resources.ToolTip_PackageRecommended}"
+                        Visibility="{Binding Recommended, Converter={StaticResource BooleanToVisibilityConverter}}"
+                        Text="&#x2605; " />
```
Please test with different VS themes. I believe the Foreground needs to use our UIText brush so that with a black/dark background theme in VS, the star will become white and visible.

birgithi replied:
I tested with Light, Dark, Blue and Blue Extra Contrast. In all cases the star matched the text color of the package name, either black or white.

donnie-msft replied:
Sounds like we have an inherited style somewhere - awesome!

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -733,10 +755,45 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
+                {
+                    if (project.ProjectStyle == ProjectModel.ProjectStyle.PackagesConfig)
+                    {
+                        _recommendPackages = true;
```
Isn't this unnecessary if there's only 1 Project (IsSolution is false)?

(And anyway, if this were checking Solution PMUI, I think this logic is wrong since any PC project in the Solution would enable recommendations for this Package Manager. Should be checking that All Projects are PC, right?)

birgithi replied:
I rearranged the order of the tests. 
You are correct, there is only one project when IsSolution is false.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -733,10 +755,45 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
+            //   this is the Browse tab,
+            //   and the search text is an empty string
+            _recommendPackages = false;
+            if (loadContext.IsSolution == false
```
Can we check for PC first before iterating Sources? Wanting to minimize perf impact.

birgithi replied:
I rearranged the order of the tests.

birgithi committed:
Cleanup for PR feedback

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -909,57 +969,87 @@ private void EmitSearchSelectionTelemetry(PackageItemListViewModel selectedPacka
+                    context.SourceRepositories,
+                    uiLogger,
+                    TelemetryActivity.NuGetTelemetryService);
+                    packageFeeds.recommenderFeed = new RecommenderPackageFeed(
```
formatting

birgithi replied:
Fixed

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -192,6 +196,33 @@ public async Task UpdateStateAsync(IProgress<IItemLoaderState> progress, Cancell
+            SearchResult<IPackageSearchMetadata> aggregated = browseResult;
+            if (recommenderResult != null)
+            {
+                IEnumerable<IPackageSearchMetadata> items = recommenderResult.Items.Concat(browseResult.Items);
```
Just for clarity in my understanding, the intention is to display the recommendation, and also display that same package as it always has been (without the *) in the list, wherever that would have normally appeared?

birgithi replied:
We decided to filter out duplicates if they appear in the first page (25 results). For example if we recommend Newtonsoft.Json, we would show it with the star, but then remove it from the non-starred list. For anything past the first 25 results, we'll just list duplicates so that we don't need to keep checking against the recomended list.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -192,6 +196,33 @@ public async Task UpdateStateAsync(IProgress<IItemLoaderState> progress, Cancell
+                _recommendedCount = recommenderResult.Count();
+            }
+
+            // update items list to include recommender results if needed
```
What I think it actually does, and would be more clear
```suggestion
            // append recommender results if needed
```

birgithi replied:
Changed to "if there are recommendations, add them to the top of the browse results list"

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -733,10 +755,45 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
+            else
+            {
+                packageFeeds = await CreatePackageFeedAsync(loadContext, _topPanel.Filter, _uiLogger, false);
+            }
```
Please extract as much of this to a separate method as possible. Would like to keep the Search method as clear and minimal as possible.

birgithi replied:
Moved this into a separate method.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -909,57 +969,87 @@ private void EmitSearchSelectionTelemetry(PackageItemListViewModel selectedPacka
             if (context.SourceRepositories == null)
             {
-                return null;
+                return packageFeeds;
```
Logic elsewhere does null checks... I believe the named Tuple values being `null` will not get treated as `null`.
Can you confirm this validation still works? https://github.com/NuGet/NuGet.Client/blob/afb506a30c5d4b2d7ecf838b198ef01858e3f398/src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs#L96

birgithi replied:
I call PackageItemLoader like this: PackageItemLoader(loadContext, packageFeeds.mainFeed, searchText, IncludePrerelease, packageFeeds.recommenderFeed), so it should use the correct values. Just to be sure I tried changing packageFeeds.mainFeed to null before the call, and the null check worked as expected.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,192 @@
+                throw new InvalidOperationException("Invalid token");
+            }
+            // don't make recommendations if the user entered a search string
+            if (searchToken.SearchString != "")
```
```suggestion
            if (!string.IsNullOrEmpty(searchToken.SearchString))
```

birgithi replied:
Fixed

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,192 @@
+            }
+
+            // get PackageIdentity info for the top 5 recommended packages
+            int _maxRecommend = 5;
```
Should this be easily configurable? At a minimum, please extract to a constant field.

birgithi replied:
Made this a constant field.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,192 @@
+            {
+                // get lists of only the package ids to send to the recommender
+                List<string> topPackages = _installedPackages.Select(item => item.Id.ToLowerInvariant()).ToList();
+                List<string> depPackages = _dependentPackages.Keys.ToList();
```
Wouldn't we want to eliminate duplicates, here? Anyway, this is all dead code since we're not implementing PR... not sure if it should all be commented out or removed or left in place.

birgithi replied:
I removed the rest of the dependent packages code. All I left is creating an empty list of package ids to send to the recommender, since the recommender still expects this parameter, and is able to use dependent packages in recommendations when available.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,192 @@
+                List<string> topPackages = _installedPackages.Select(item => item.Id.ToLowerInvariant()).ToList();
+                List<string> depPackages = _dependentPackages.Keys.ToList();
+                // call the recommender to get package recommendations
+                recommendIds = await NuGetRecommender.GetRecommendedPackageIdsAsync(_targetFrameworks, topPackages, depPackages, cancellationToken);
```
Seems to me we should be able to pass the Max Recommended to this service so that it doesn't send us possibly way more than necessary.

birgithi replied:
Makes sense to not pass back more than what we need, although I can envision scenarios where that might not match the number we show in the list. I'll update the recommender to accept a max as input, then come back to update it here.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,192 @@
+            int _maxRecommend = 5;
+            int index = 0;
+            List<PackageIdentity> recommendPackages = new List<PackageIdentity>();
+            while (recommendIds != null && index < recommendIds.Count() && recommendPackages.Count < _maxRecommend)
```
If `recommendIds` is null, a whole lot of the following logic is irrelevant. I would prefer an if-statement for this null check that wraps what's relevant below this point.

birgithi replied:
Rearranged the code so that this is only used when there are recommendations.

donnie-msft reviewed (commented):
Nice work... I dug a little deeper this time around

birgithi committed:
Changes due to rebase, filter duplicates of recommended packages

birgithi committed:
Code cleanup for PR feedback

birgithi committed:
More PR feedback changes

birgithi committed:
Remove rest of dependent packages code

birgithi committed:
Use AsyncLazy to get recommender service

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -192,6 +196,36 @@ public async Task UpdateStateAsync(IProgress<IItemLoaderState> progress, Cancell
+            SearchResult<IPackageSearchMetadata> browseResult = await _packageFeed.SearchAsync(_searchText, SearchFilter, cancellationToken);
+
+            // get the recommender results
+            SearchResult<IPackageSearchMetadata> recommenderResult = null;
```
nit: In the future, these should probably happen in parallel.

nkolev92 replied:
Specifically serach and recommender query...a new issues might be good for that.

birgithi replied:
Created issue https://github.com/NuGet/Home/issues/9572

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -719,6 +736,48 @@ private void SearchPackagesAndRefreshUpdateCount(bool useCacheForUpdates)
+                && _topPanel.Filter == ItemFilter.All
+                && searchText == string.Empty
+                && loadContext.SourceRepositories.Count() == 1
+                && Model.Context.Projects.Count() == 1
```
this is unnecessary with the IsSolution check. 
No model that's not solution based should contain more than 1 project.

birgithi replied:
removed

nkolev92 reviewed (approved):
I think it's fine for the most part.

Especially given that we want to make some changes in future PRs.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -719,6 +736,48 @@ private void SearchPackagesAndRefreshUpdateCount(bool useCacheForUpdates)
+            }
+            else
+            {
+                packageFeeds = await CreatePackageFeedAsync(loadContext, _topPanel.Filter, _uiLogger, false);
```
```suggestion
                packageFeeds = await CreatePackageFeedAsync(loadContext, _topPanel.Filter, _uiLogger, recommendPackages: false);
```

birgithi replied:
Fixed

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -98,6 +103,8 @@ internal PackageSourceMoniker SelectedSource
         private readonly Guid _sessionGuid = Guid.NewGuid();
         private readonly Stopwatch _sinceLastRefresh;
 
+        private bool _forceRecommender = false;
```
default value unnecessary
```suggestion
        private bool _forceRecommender;
```

birgithi replied:
Fixed

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -793,9 +852,9 @@ private void RefreshInstalledAndUpdatesTabs()
                 _topPanel.UpdateCountOnUpdatesTab(count: 0);
                 var loadContext = new PackageLoadContext(ActiveSources, Model.IsSolution, Model.Context);
-                var packageFeed = await CreatePackageFeedAsync(loadContext, ItemFilter.UpdatesAvailable, _uiLogger);
+                var packageFeeds = await CreatePackageFeedAsync(loadContext, ItemFilter.UpdatesAvailable, _uiLogger, false);
```
```suggestion
                var packageFeeds = await CreatePackageFeedAsync(loadContext, ItemFilter.UpdatesAvailable, _uiLogger, recommendPackages: false);
```

birgithi replied:
Fixed

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -856,9 +915,9 @@ private void RefreshConsolidatablePackagesCount()
                     _topPanel.UpdateCountOnConsolidateTab(count: 0);
                     var loadContext = new PackageLoadContext(ActiveSources, Model.IsSolution, Model.Context);
-                    var packageFeed = await CreatePackageFeedAsync(loadContext, ItemFilter.Consolidate, _uiLogger);
+                    var packageFeeds = await CreatePackageFeedAsync(loadContext, ItemFilter.Consolidate, _uiLogger, false);
```
```suggestion
                    var packageFeeds = await CreatePackageFeedAsync(loadContext, ItemFilter.Consolidate, _uiLogger, recommendPackages: false);
```

birgithi replied:
Fixed

donnie-msft reviewed (approved):
Thanks for addressing everything. Just a couple more nits. 👍

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLoadContext.cs:
```diff
@@ -3,13 +3,18 @@
+using NuGet.ProjectManagement.Projects;
+using NuGet.ProjectModel;
 using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
```
This file was changed to add a bunch of usings, but there is no code changes, so these usings must be unused. It'd be great if they could be cleaned up.

birgithi replied:
They were left from previous code changes. I have removed them.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -184,6 +191,16 @@ internal PackageSourceMoniker SelectedSource
+            // check if environment variable RecommendNuGetPackages to turn on recommendations is set to 1
+            try
+            {
+                _forceRecommender = (Environment.GetEnvironmentVariable("RecommendNuGetPackages") == "1");
```
probably a nitpick to fix in another PR, but I think NuGet already uses some environment variablies starting with the `NUGET_` prefix. If that's right, we should use the same convention.

birgithi replied:
I changed it to NUGET_RECOMMEND_PACKAGES.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,200 @@
+            IPackageMetadataProvider metadataProvider,
+            Common.ILogger logger)
+        {
+            if (sourceRepository == null)
+            {
+                throw new ArgumentNullException(nameof(sourceRepository));
+            }
+            _sourceRepository = sourceRepository;
+
+            if (installedPackages == null)
+            {
+                throw new ArgumentNullException(nameof(installedPackages));
+            }
+            _installedPackages = installedPackages;
+
+            if (targetFrameworks == null)
+            {
+                throw new ArgumentNullException(nameof(targetFrameworks));
+            }
+            _targetFrameworks = targetFrameworks;
+
+            if (metadataProvider == null)
+            {
+                throw new ArgumentNullException(nameof(metadataProvider));
+            }
+            _metadataProvider = metadataProvider;
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            _logger = logger;
```
```suggestion
            _sourceRepository = sourceRepository ?? throw new ArgumentNullException(nameof(sourceRepository));
            _installedPackages = installedPackages ?? throw new ArgumentNullException(nameof(installedPackages));
            _targetFrameworks = targetFrameworks ?? throw new ArgumentNullException(nameof(targetFrameworks));
            _metadataProvider = metadataProvider ?? throw new ArgumentNullException(nameof(metadataProvider));
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
```

birgithi replied:
Fixed

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,200 @@
+            var searchToken = continuationToken as RecommendSearchToken;
+            if (searchToken == null)
+            {
+                throw new InvalidOperationException("Invalid token");
```
```suggestion
                throw new ArgumentException("Invalid continuation token", nameof(continuationToken));
```

birgithi replied:
Fixed

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,200 @@
+            // get PackageIdentity info for the top 5 recommended packages
+            int index = 0;
+            List<PackageIdentity> recommendPackages = new List<PackageIdentity>();
+            while (index < recommendIds.Count() && recommendPackages.Count < MaxRecommended)
```
```suggestion
            while (index < recommendIds.Count && recommendPackages.Count < MaxRecommended)
```

birgithi replied:
Fixed

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs:
```diff
@@ -0,0 +1,200 @@
+                MetadataResource _metadataResource = await _sourceRepository.GetResourceAsync<MetadataResource>(cancellationToken);
+                PackageMetadataResource _packageMetadataResource = await _sourceRepository.GetResourceAsync<PackageMetadataResource>(cancellationToken);
+
+                Versioning.NuGetVersion ver = await _metadataResource.GetLatestVersion(recommendIds[index], false, false, NullSourceCacheContext.Instance, Common.NullLogger.Instance, cancellationToken);
```
nitpick: use named parameters for the two `false` parameters.

birgithi replied:
Fixed

zivkan reviewed (approved):
I only have nitpicks.

birgithi committed:
PR feedback changes
